components/common/src/command/package/install.rs

```diff
@@ -502,43 +625,139 @@ impl<'a> InstallTask<'a> {
+                if pkg_install.as_ref() > pkg_artifact.as_ref() {
+                    pkg_install
+                } else {
+                    pkg_artifact
+                }
+            }
+            (Err(_), Err(_)) => return Err(Error::PackageNotFound),
+        };
+        debug!("offline mode: winner: {:?}", &latest);
+
+        Ok(latest)
     }
 
-    // TODO (CM): This could return a plain bool IF we could ensure
-    // above that the package identifier is FULLY QUALIFIED
-    fn is_artifact_cached(&self, ident: &PackageIdent) -> Result<bool> {
-        Ok(self.cached_artifact_path(ident)?.is_file())
+    fn latest_installed_ident(&self, ident: &PackageIdent) -> Result<FullyQualifiedPackageIdent> {
+        match PackageInstall::load(ident, Some(self.fs_root_path)) {
+            Ok(pi) => FullyQualifiedPackageIdent::from(pi.ident().clone()),
+            Err(_) => Err(Error::PackageNotFound),
+        }
+    }
+
+    fn latest_cached_ident(&self, ident: &PackageIdent) -> Result<FullyQualifiedPackageIdent> {
+        let filename_glob = {
+            let mut ident = ident.clone();
+            if ident.version.is_none() {
+                ident.version = Some(String::from("_VERSION_"));
+            }
+            if ident.release.is_none() {
+                ident.release = Some(String::from("_RELEASE_"));
+            }
+            match ident.archive_name() {
+                Some(name) => name.replace("_VERSION_", "?*").replace("_RELEASE_", "?*"),
+                None => return Err(Error::PackageNotFound),
+            }
+        };
+        let glob_path = self.artifact_cache_path.join(filename_glob);
+        let glob_path = glob_path.to_string_lossy();
+        debug!("looking for cached artifacts, glob={}", glob_path.as_ref());
+
+        let mut latest: Vec<(PackageIdent, PackageArchive)> = Vec::with_capacity(1);
+        for file in glob::glob(glob_path.as_ref())
+            .expect("glob pattern should compile")
+            .filter_map(StdResult::ok)
+        {
+            let mut artifact = PackageArchive::new(&file);
+            let artifact_ident = artifact.ident().ok();
+            if let None = artifact_ident {
+                continue;
+            }
+            let artifact_ident = artifact_ident.unwrap();
+            if artifact_ident.origin == ident.origin && artifact_ident.name == ident.name {
+                if latest.is_empty() {
+                    latest.push((artifact_ident, artifact));
+                } else {
+                    if artifact_ident > latest[0].0 {
+                        let _ = latest.pop();
+                        latest.push((artifact_ident, artifact));
+                    }
+                }
+            }
+        }
+
+        if latest.is_empty() {
+            Err(Error::PackageNotFound)
+        } else {
+            Ok(FullyQualifiedPackageIdent::from(
+                latest.pop().unwrap().1.ident()?,
+            )?)
+        }
```

Comment:
Related to [the previous comment](https://github.com/habitat-sh/habitat/pull/4771/files#r175818297), an `Option` would allow you to avoid some of the `pop().unrwap()`-type stuff:
```rust
        latest.map_or(Err(Error::PackageNotFound), |mut l| {
            Ok(FullyQualifiedPackageIdent::from(l.1.ident()?)?)
        })
```
though, if you find the `if`/`else` easier to read, that still saves the `pop()`:
```rust
        if latest.is_none() {
            Err(Error::PackageNotFound)
        } else {
            Ok(FullyQualifiedPackageIdent::from(
                latest.unwrap().1.ident()?,
            )?)
        }
```

Reply:
Oooh, I like that.